diff --git a/aapg/utils.py b/aapg/utils.py
index 60c6742..5b1705f 100644
--- a/aapg/utils.py
+++ b/aapg/utils.py
@@ -25,7 +25,7 @@ class ColoredFormatter(logging.Formatter):
         level_name = str(record.levelname)
         name = str(record.name)
         color_prefix = self.colors[level_name]
-        return '{0}{1:<9s} - {2:<25s} : {3}{4}'.format(
+        return '{0}{1:<9s} : {3}{4}'.format(
                 color_prefix,
                 '[' + level_name + ']',
                 name,
