{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.20.3'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import h5py\n",
    "import numpy as np\n",
    "import bcolz\n",
    "import zarr\n",
    "import sys\n",
    "sys.path.insert(0, '../..')\n",
    "import allel\n",
    "allel.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# data from http://www.malariagen.net/data/ag1000g-phase1-ar3\n",
    "callset = h5py.File('/data/coluzzi/ag1000g/data/phase1/release/AR3/variation/main/hdf5/ag1000g.phase1.ar3.pass.h5',\n",
    "                    mode='r')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Chunked arrays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>GenotypeChunkedArray((9643193, 765, 2), int8, chunks=(6553, 10, 2))<br/>  nbytes: 13.7G; cbytes: 548.0M; cratio: 25.7;<br/>  compression: gzip; compression_opts: 3;<br/>  data: h5py._hl.dataset.Dataset</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>760</th>\n",
       "<th>761</th>\n",
       "<th>762</th>\n",
       "<th>763</th>\n",
       "<th>764</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "GenotypeChunkedArray((9643193, 765, 2), int8, chunks=(6553, 10, 2))\n",
       "  nbytes: 13.7G; cbytes: 548.0M; cratio: 25.7;\n",
       "  compression: gzip; compression_opts: 3;\n",
       "  data: h5py._hl.dataset.Dataset"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype = allel.GenotypeChunkedArray(callset['3L/calldata/genotype'])\n",
    "genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GenotypeChunkedArray((9643193, 765, 2), int8, chunks=(6553, 10, 2))\n",
      "  nbytes: 13.7G; cbytes: 548.0M; cratio: 25.7;\n",
      "  compression: gzip; compression_opts: 3;\n",
      "  data: h5py._hl.dataset.Dataset\n"
     ]
    }
   ],
   "source": [
    "print(repr(genotype))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 38.7 s, sys: 124 ms, total: 38.8 s\n",
      "Wall time: 38.8 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zarr.blosc.set_nthreads(8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bcolz.set_nthreads(8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2))<br/>  nbytes: 1.4G; cbytes: 85.0M; cratio: 17.2;<br/>  compression: blosc; compression_opts: cparams(clevel=5, shuffle=1, cname='lz4', quantize=0);<br/>  data: bcolz.carray_ext.carray</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>760</th>\n",
       "<th>761</th>\n",
       "<th>762</th>\n",
       "<th>763</th>\n",
       "<th>764</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2))\n",
       "  nbytes: 1.4G; cbytes: 85.0M; cratio: 17.2;\n",
       "  compression: blosc; compression_opts: cparams(clevel=5, shuffle=1, cname='lz4', quantize=0);\n",
       "  data: bcolz.carray_ext.carray"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype.copy(stop=1000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7.4 s, sys: 144 ms, total: 7.54 s\n",
      "Wall time: 2.63 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2))\n",
       "  nbytes: 1.4G; cbytes: 85.0M; cratio: 17.2;\n",
       "  compression: blosc; compression_opts: cparams(clevel=5, shuffle=1, cname='lz4', quantize=0);\n",
       "  data: bcolz.carray_ext.carray</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>760</th>\n",
       "<th>761</th>\n",
       "<th>762</th>\n",
       "<th>763</th>\n",
       "<th>764</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2))\n",
       "  nbytes: 1.4G; cbytes: 85.0M; cratio: 17.2;\n",
       "  compression: blosc; compression_opts: cparams(clevel=5, shuffle=1, cname='lz4', quantize=0);\n",
       "  data: bcolz.carray_ext.carray"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_bcolzmem = genotype.copy(stop=1000000, storage='bcolzmem')\n",
    "%time genotype_bcolzmem.max()\n",
    "genotype_bcolzmem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7.49 s, sys: 64 ms, total: 7.56 s\n",
      "Wall time: 2.73 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2))\n",
       "  nbytes: 1.4G; cbytes: 83.1M; cratio: 17.6;\n",
       "  compression: blosc; compression_opts: {'clevel': 5, 'shuffle': 1, 'cname': 'lz4'};\n",
       "  data: zarr.core.Array</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>760</th>\n",
       "<th>761</th>\n",
       "<th>762</th>\n",
       "<th>763</th>\n",
       "<th>764</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2))\n",
       "  nbytes: 1.4G; cbytes: 83.1M; cratio: 17.6;\n",
       "  compression: blosc; compression_opts: {'clevel': 5, 'shuffle': 1, 'cname': 'lz4'};\n",
       "  data: zarr.core.Array"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_zarrmem = genotype.copy(stop=1000000, storage='zarrmem', chunks=genotype_bcolzmem.chunks,\n",
    "                                 compression_opts=dict(cname='lz4', clevel=5, shuffle=1))\n",
    "%time genotype_zarrmem.max()\n",
    "genotype_zarrmem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6.24 s, sys: 72 ms, total: 6.31 s\n",
      "Wall time: 2.47 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2), nbytes=1.4G, cbytes=85.0M, cratio=17.2, compression=blosc, compression_opts=cparams(clevel=5, shuffle=1, cname='lz4', quantize=0), data=bcolz.carray_ext.carray)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>760</th>\n",
       "<th>761</th>\n",
       "<th>762</th>\n",
       "<th>763</th>\n",
       "<th>764</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2), nbytes=1.4G, cbytes=85.0M, cratio=17.2, compression=blosc, compression_opts=cparams(clevel=5, shuffle=1, cname='lz4', quantize=0), data=bcolz.carray_ext.carray)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_bcolztmp = genotype.copy(stop=1000000, storage='bcolztmp')\n",
    "%time genotype_bcolztmp.max()\n",
    "genotype_bcolztmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 9.48 s, sys: 164 ms, total: 9.65 s\n",
      "Wall time: 3.09 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2), nbytes=1.4G, cbytes=54.7M, cratio=26.7, compression=blosc, compression_opts=cparams(clevel=1, shuffle=1, cname='zlib', quantize=0), data=bcolz.carray_ext.carray)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>760</th>\n",
       "<th>761</th>\n",
       "<th>762</th>\n",
       "<th>763</th>\n",
       "<th>764</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(1370, 765, 2), nbytes=1.4G, cbytes=54.7M, cratio=26.7, compression=blosc, compression_opts=cparams(clevel=1, shuffle=1, cname='zlib', quantize=0), data=bcolz.carray_ext.carray)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_bcolzmem_zlib1 = genotype.copy(stop=1000000, storage='bcolzmem_zlib1')\n",
    "%time genotype_bcolzmem_zlib1.max()\n",
    "genotype_bcolzmem_zlib1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 9.88 s, sys: 240 ms, total: 10.1 s\n",
      "Wall time: 3.16 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_bcolztmp_zlib1 = genotype.copy(stop=1000000, storage='bcolztmp_zlib1')\n",
    "%time genotype_bcolztmp_zlib1.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.12 s, sys: 4 ms, total: 4.12 s\n",
      "Wall time: 4.12 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(685, 765, 2), nbytes=1.4G, cbytes=52.4M, cratio=27.8, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>760</th>\n",
       "<th>761</th>\n",
       "<th>762</th>\n",
       "<th>763</th>\n",
       "<th>764</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "GenotypeChunkedArray((1000000, 765, 2), int8, chunks=(685, 765, 2), nbytes=1.4G, cbytes=52.4M, cratio=27.8, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_hdf5mem_zlib1 = genotype.copy(stop=1000000, storage='hdf5mem_zlib1')\n",
    "%time genotype_hdf5mem_zlib1.max()\n",
    "genotype_hdf5mem_zlib1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.05 s, sys: 36 ms, total: 4.09 s\n",
      "Wall time: 4.08 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_hdf5tmp_zlib1 = genotype.copy(stop=1000000, storage='hdf5tmp_zlib1')\n",
    "%time genotype_hdf5tmp_zlib1.max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tune default HDF5 chunk size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2741"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype.copy(storage='bcolzmem').chunklen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1370"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_bcolzmem.chunklen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(685, 765, 2)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_hdf5mem_zlib1.chunks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32768 21\n",
      "CPU times: user 9.11 s, sys: 0 ns, total: 9.11 s\n",
      "Wall time: 9.09 s\n",
      "65536 42\n",
      "CPU times: user 5.96 s, sys: 0 ns, total: 5.96 s\n",
      "Wall time: 5.95 s\n",
      "131072 85\n",
      "CPU times: user 4.39 s, sys: 0 ns, total: 4.39 s\n",
      "Wall time: 4.39 s\n",
      "262144 171\n",
      "CPU times: user 3.61 s, sys: 0 ns, total: 3.61 s\n",
      "Wall time: 3.6 s\n",
      "524288 342\n",
      "CPU times: user 3.19 s, sys: 0 ns, total: 3.19 s\n",
      "Wall time: 3.18 s\n",
      "1048576 685\n",
      "CPU times: user 3.78 s, sys: 0 ns, total: 3.78 s\n",
      "Wall time: 3.77 s\n",
      "2097152 1370\n",
      "CPU times: user 2.98 s, sys: 0 ns, total: 2.98 s\n",
      "Wall time: 2.97 s\n",
      "4194304 2741\n",
      "CPU times: user 2.89 s, sys: 0 ns, total: 2.89 s\n",
      "Wall time: 2.89 s\n",
      "8388608 5482\n",
      "CPU times: user 2.89 s, sys: 0 ns, total: 2.89 s\n",
      "Wall time: 2.89 s\n",
      "16777216 10965\n",
      "CPU times: user 2.98 s, sys: 0 ns, total: 2.98 s\n",
      "Wall time: 2.98 s\n",
      "33554432 21931\n",
      "CPU times: user 2.99 s, sys: 140 ms, total: 3.13 s\n",
      "Wall time: 3.13 s\n"
     ]
    }
   ],
   "source": [
    "for chunksize in 2**15, 2**16, 2**17, 2**18, 2**19, 2**20, 2**21, 2**22, 2**23, 2**24, 2**25:\n",
    "    chunklen = chunksize // (genotype.shape[1] * genotype.shape[2])\n",
    "    chunks = (chunklen,) + genotype.shape[1:]\n",
    "    print(chunksize, chunklen)\n",
    "    genotype_hdf5mem_zlib1_bigchunks = genotype.copy(stop=1000000, storage='hdf5mem_zlib1', chunks=chunks)\n",
    "    %time genotype_hdf5mem_zlib1_bigchunks.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32768 21\n",
      "CPU times: user 4.81 s, sys: 0 ns, total: 4.81 s\n",
      "Wall time: 4.81 s\n",
      "65536 42\n",
      "CPU times: user 3.33 s, sys: 4 ms, total: 3.33 s\n",
      "Wall time: 3.32 s\n",
      "131072 85\n",
      "CPU times: user 7.37 s, sys: 724 ms, total: 8.1 s\n",
      "Wall time: 4.05 s\n",
      "262144 171\n",
      "CPU times: user 7.29 s, sys: 432 ms, total: 7.72 s\n",
      "Wall time: 3.2 s\n",
      "524288 342\n",
      "CPU times: user 7.7 s, sys: 276 ms, total: 7.97 s\n",
      "Wall time: 3.15 s\n",
      "1048576 685\n",
      "CPU times: user 7.26 s, sys: 92 ms, total: 7.35 s\n",
      "Wall time: 2.58 s\n",
      "2097152 1370\n",
      "CPU times: user 6.6 s, sys: 80 ms, total: 6.68 s\n",
      "Wall time: 2.32 s\n",
      "4194304 2741\n",
      "CPU times: user 6.16 s, sys: 24 ms, total: 6.18 s\n",
      "Wall time: 2.21 s\n",
      "8388608 5482\n",
      "CPU times: user 5.48 s, sys: 4 ms, total: 5.48 s\n",
      "Wall time: 1.99 s\n",
      "16777216 10965\n",
      "CPU times: user 5.12 s, sys: 8 ms, total: 5.13 s\n",
      "Wall time: 1.8 s\n",
      "33554432 21931\n",
      "CPU times: user 4.53 s, sys: 480 ms, total: 5.01 s\n",
      "Wall time: 1.55 s\n"
     ]
    }
   ],
   "source": [
    "for chunksize in 2**15, 2**16, 2**17, 2**18, 2**19, 2**20, 2**21, 2**22, 2**23, 2**24, 2**25:\n",
    "    chunklen = chunksize // (genotype.shape[1] * genotype.shape[2])\n",
    "    print(chunksize, chunklen)\n",
    "    genotype_bcolzmem_bigchunks = genotype.copy(stop=1000000, storage='bcolzmem', chunklen=chunklen)\n",
    "    %time genotype_bcolzmem_bigchunks.max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Excercise methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 12.1 s, sys: 160 ms, total: 12.3 s\n",
      "Wall time: 3.04 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "zarr.core.Array((1000000, 765), bool, chunks=(1370, 765), order=C)\n",
       "  compression: blosc; compression_opts: {'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}\n",
       "  nbytes: 729.6M; nbytes_stored: 3.6M; ratio: 204.7; initialized: 730/730\n",
       "  store: builtins.dict"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_bcolzmem.is_called()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 10.6 s, sys: 180 ms, total: 10.8 s\n",
      "Wall time: 6.98 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<HDF5 dataset \"data\": shape (1000000, 765), type \"|b1\">"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_hdf5mem_zlib1.is_called(storage='hdf5mem_zlib1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 14.1 s, sys: 176 ms, total: 14.3 s\n",
      "Wall time: 3.45 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "28029171"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_bcolzmem.count_het()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11.1 s, sys: 164 ms, total: 11.3 s\n",
      "Wall time: 5.44 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "28029171"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_hdf5mem_zlib1.count_het()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 17.1 s, sys: 292 ms, total: 17.4 s\n",
      "Wall time: 7.52 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=4.3M, cratio=3.6, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>1527</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>1529</td>\n",
       "<td>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>1526</td>\n",
       "<td>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=4.3M, cratio=3.6, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_bcolzmem.count_alleles()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 23.8 s, sys: 580 ms, total: 24.4 s\n",
      "Wall time: 8.81 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=4.3M, cratio=3.6, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>1527</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>1529</td>\n",
       "<td>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>1526</td>\n",
       "<td>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=4.3M, cratio=3.6, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_bcolzmem_zlib1.count_alleles()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7.89 s, sys: 4 ms, total: 7.89 s\n",
      "Wall time: 7.88 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=2.4M, cratio=6.3, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>1527</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>1529</td>\n",
       "<td>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>1526</td>\n",
       "<td>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=2.4M, cratio=6.3, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_hdf5mem_zlib1.count_alleles(storage='hdf5mem_zlib1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedTable(1000000, nbytes=30.5M, cbytes=4.3M, cratio=7.1, data=allel.chunked.storage_zarr.ZarrTable)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>pop1</th>\n",
       "<th>pop2</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>[200   0   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>[199   1   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>[200   0   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>[198   2   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>[196   4   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedTable(1000000, nbytes=30.5M, cbytes=4.3M, cratio=7.1, data=allel.chunked.storage_zarr.ZarrTable)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acs = genotype_bcolzmem.count_alleles_subpops(subpops={'pop1': list(range(100)),\n",
    "                                                       'pop2': list(range(100, 200))},\n",
    "                                              max_allele=3)\n",
    "acs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=2.2M, cratio=6.9, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>200</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>199</td>\n",
       "<td>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>200</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>198</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>196</td>\n",
       "<td>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=2.2M, cratio=6.9, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acs['pop1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedTable(1000000, nbytes=30.5M, cbytes=3.0M, cratio=10.2, data=h5py._hl.files.File)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>pop1</th>\n",
       "<th>pop2</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>[200   0   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>[199   1   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>[200   0   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>[198   2   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>[196   4   0   0]</td>\n",
       "<td>[200   0   0   0]</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedTable(1000000, nbytes=30.5M, cbytes=3.0M, cratio=10.2, data=h5py._hl.files.File)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acs = genotype_hdf5mem_zlib1.count_alleles_subpops(subpops={'pop1': list(range(100)),\n",
    "                                                            'pop2': list(range(100, 200))},\n",
    "                                                   max_allele=3,\n",
    "                                                   storage='hdf5mem_zlib1')\n",
    "acs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=1.5M, cratio=10.2, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>200</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>199</td>\n",
       "<td>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>200</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>198</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>196</td>\n",
       "<td>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=1.5M, cratio=10.2, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acs['pop1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 10.3 s, sys: 264 ms, total: 10.5 s\n",
      "Wall time: 3.43 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>HaplotypeChunkedArray((1000000, 1530), int8, chunks=(685, 1530), nbytes=1.4G, cbytes=78.5M, cratio=18.6, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>1525</th>\n",
       "<th>1526</th>\n",
       "<th>1527</th>\n",
       "<th>1528</th>\n",
       "<th>1529</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "HaplotypeChunkedArray((1000000, 1530), int8, chunks=(685, 1530), nbytes=1.4G, cbytes=78.5M, cratio=18.6, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_bcolzmem.to_haplotypes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 19.9 s, sys: 1.14 s, total: 21 s\n",
      "Wall time: 4.06 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>HaplotypeChunkedArray((1000000, 1530), int8, chunks=(1370, 1530), nbytes=1.4G, cbytes=54.7M, cratio=26.7, compression=blosc, compression_opts=cparams(clevel=1, shuffle=1, cname='zlib', quantize=0), data=bcolz.carray_ext.carray)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>1525</th>\n",
       "<th>1526</th>\n",
       "<th>1527</th>\n",
       "<th>1528</th>\n",
       "<th>1529</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "HaplotypeChunkedArray((1000000, 1530), int8, chunks=(1370, 1530), nbytes=1.4G, cbytes=54.7M, cratio=26.7, compression=blosc, compression_opts=cparams(clevel=1, shuffle=1, cname='zlib', quantize=0), data=bcolz.carray_ext.carray)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_bcolzmem_zlib1.to_haplotypes(storage='bcolzmem_zlib1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 7.9 s, sys: 64 ms, total: 7.96 s\n",
      "Wall time: 7.96 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>HaplotypeChunkedArray((1000000, 1530), int8, chunks=(685, 1530), nbytes=1.4G, cbytes=52.4M, cratio=27.8, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>1525</th>\n",
       "<th>1526</th>\n",
       "<th>1527</th>\n",
       "<th>1528</th>\n",
       "<th>1529</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>...</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "HaplotypeChunkedArray((1000000, 1530), int8, chunks=(685, 1530), nbytes=1.4G, cbytes=52.4M, cratio=27.8, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time genotype_hdf5mem_zlib1.to_haplotypes(storage='hdf5mem_zlib1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=4.3M, cratio=3.6, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>1527</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>1529</td>\n",
       "<td>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>1526</td>\n",
       "<td>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedArray((1000000, 4), int32, chunks=(65536, 4), nbytes=15.3M, cbytes=4.3M, cratio=3.6, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ac = genotype_bcolzmem.count_alleles()\n",
    "ac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChunkedArray((1000000,), bool, chunks=(1048576,), nbytes=976.6K, cbytes=6.3K, cratio=154.7, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ac.is_segregating()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChunkedArray((1000000,), int8, chunks=(1048576,), nbytes=976.6K, cbytes=229.8K, cratio=4.2, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ac.max_allele()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Copy into persistent HDF5 file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "h5f = h5py.File('test.h5', mode='w')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 group \"/calldata\" (0 members)>"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5g = h5f.require_group('calldata')\n",
    "h5g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "if 'genotype' in h5g:\n",
    "    del h5g['genotype']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>GenotypeChunkedArray((100000, 765, 2), int8, chunks=(685, 765, 2), nbytes=145.9M, cbytes=5.0M, cratio=29.0, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "<th>4</th>\n",
       "<th>...</th>\n",
       "<th>760</th>\n",
       "<th>761</th>\n",
       "<th>762</th>\n",
       "<th>763</th>\n",
       "<th>764</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>...</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "<td>0/0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "GenotypeChunkedArray((100000, 765, 2), int8, chunks=(685, 765, 2), nbytes=145.9M, cbytes=5.0M, cratio=29.0, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc = genotype.copy(stop=100000, storage='hdf5', group=h5g, name='genotype', compression='gzip', compression_opts=1)\n",
    "gc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 dataset \"genotype\": shape (100000, 765, 2), type \"|i1\">"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5g['genotype']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>AlleleCountsChunkedArray((100000, 4), int32, chunks=(65536, 4), nbytes=1.5M, cbytes=229.0K, cratio=6.8, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th></th>\n",
       "<th>0</th>\n",
       "<th>1</th>\n",
       "<th>2</th>\n",
       "<th>3</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>0</td>\n",
       "<td>1527</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>1</td>\n",
       "<td>1529</td>\n",
       "<td>1</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>2</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>3</td>\n",
       "<td>1528</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='font-weight: bold'>4</td>\n",
       "<td>1526</td>\n",
       "<td>4</td>\n",
       "<td>0</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "AlleleCountsChunkedArray((100000, 4), int32, chunks=(65536, 4), nbytes=1.5M, cbytes=229.0K, cratio=6.8, compression=gzip, compression_opts=1, data=h5py._hl.dataset.Dataset)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ac = gc.count_alleles(storage='hdf5', group=h5g, name='allele_counts', compression='gzip', compression_opts=1)\n",
    "ac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 dataset \"allele_counts\": shape (100000, 4), type \"<i4\">"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5g['allele_counts']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Chunked tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>VariantChunkedTable(9643193, nbytes=3.7G, cbytes=424.9M, cratio=8.8, data=h5py._hl.group.Group)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>ABHet</th>\n",
       "<th>ABHom</th>\n",
       "<th>AC</th>\n",
       "<th>AF</th>\n",
       "<th>ALT</th>\n",
       "<th>AN</th>\n",
       "<th>ANN</th>\n",
       "<th>Accessible</th>\n",
       "<th>BaseCounts</th>\n",
       "<th>BaseQRankSum</th>\n",
       "<th>CHROM</th>\n",
       "<th>Coverage</th>\n",
       "<th>CoverageMQ0</th>\n",
       "<th>DP</th>\n",
       "<th>DS</th>\n",
       "<th>Dels</th>\n",
       "<th>FILTER_FS</th>\n",
       "<th>FILTER_HRun</th>\n",
       "<th>FILTER_HighCoverage</th>\n",
       "<th>FILTER_HighMQ0</th>\n",
       "<th>FILTER_LowCoverage</th>\n",
       "<th>FILTER_LowMQ</th>\n",
       "<th>FILTER_LowQual</th>\n",
       "<th>FILTER_NoCoverage</th>\n",
       "<th>FILTER_PASS</th>\n",
       "<th>FILTER_QD</th>\n",
       "<th>FILTER_ReadPosRankSum</th>\n",
       "<th>FILTER_RefN</th>\n",
       "<th>FILTER_RepeatDUST</th>\n",
       "<th>FS</th>\n",
       "<th>HRun</th>\n",
       "<th>HW</th>\n",
       "<th>HaplotypeScore</th>\n",
       "<th>HighCoverage</th>\n",
       "<th>HighMQ0</th>\n",
       "<th>InbreedingCoeff</th>\n",
       "<th>LOF</th>\n",
       "<th>LowCoverage</th>\n",
       "<th>LowMQ</th>\n",
       "<th>LowPairing</th>\n",
       "<th>MLEAC</th>\n",
       "<th>MLEAF</th>\n",
       "<th>MQ</th>\n",
       "<th>MQ0</th>\n",
       "<th>MQRankSum</th>\n",
       "<th>NDA</th>\n",
       "<th>NMD</th>\n",
       "<th>NoCoverage</th>\n",
       "<th>OND</th>\n",
       "<th>POS</th>\n",
       "<th>QD</th>\n",
       "<th>QUAL</th>\n",
       "<th>REF</th>\n",
       "<th>RPA</th>\n",
       "<th>RU</th>\n",
       "<th>ReadPosRankSum</th>\n",
       "<th>RefMasked</th>\n",
       "<th>RefN</th>\n",
       "<th>RepeatDUST</th>\n",
       "<th>RepeatMasker</th>\n",
       "<th>RepeatTRF</th>\n",
       "<th>STR</th>\n",
       "<th>VariantType</th>\n",
       "<th>is_snp</th>\n",
       "<th>num_alleles</th>\n",
       "<th>svlen</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>0.50195</td>\n",
       "<td>0.99805</td>\n",
       "<td>[3 0 0]</td>\n",
       "<td>[ 0.00196075  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   21 25251    11    77]</td>\n",
       "<td>-1.8486</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25889</td>\n",
       "<td>41</td>\n",
       "<td>23344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.88184</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0332</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>-0.0025005</td>\n",
       "<td>b''</td>\n",
       "<td>18</td>\n",
       "<td>1</td>\n",
       "<td>368</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>42</td>\n",
       "<td>0.30811</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.002327</td>\n",
       "<td>9790</td>\n",
       "<td>14.047</td>\n",
       "<td>1447.12</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.22998</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[1 0 0]</td>\n",
       "<td>[ 0.00065374  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   39     9 25274    18]</td>\n",
       "<td>-3.5195</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25916</td>\n",
       "<td>14</td>\n",
       "<td>23321</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.4609</td>\n",
       "<td>0</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0449</td>\n",
       "<td>5</td>\n",
       "<td>0</td>\n",
       "<td>-0.0012999</td>\n",
       "<td>b''</td>\n",
       "<td>15</td>\n",
       "<td>1</td>\n",
       "<td>370</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>15</td>\n",
       "<td>3.2109</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017834</td>\n",
       "<td>9798</td>\n",
       "<td>13.781</td>\n",
       "<td>551.05</td>\n",
       "<td>b'G'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.13501</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.52783</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   45 25873    13    16]</td>\n",
       "<td>-3.0566</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26223</td>\n",
       "<td>8</td>\n",
       "<td>23877</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.3818</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0645</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>23</td>\n",
       "<td>0</td>\n",
       "<td>300</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.094</td>\n",
       "<td>8</td>\n",
       "<td>2.498</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018845</td>\n",
       "<td>9812</td>\n",
       "<td>13.273</td>\n",
       "<td>729.69</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.077026</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   43    21     9 26227]</td>\n",
       "<td>0.18298</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26516</td>\n",
       "<td>8</td>\n",
       "<td>24217</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.94775</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>19</td>\n",
       "<td>0</td>\n",
       "<td>279</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.156</td>\n",
       "<td>8</td>\n",
       "<td>3.2871</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017052</td>\n",
       "<td>9818</td>\n",
       "<td>14.352</td>\n",
       "<td>1147.7</td>\n",
       "<td>b'T'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.68213</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.48804</td>\n",
       "<td>0.99805</td>\n",
       "<td>[4 0 0]</td>\n",
       "<td>[ 0.00261307  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   14 26351     2   129]</td>\n",
       "<td>-3.1367</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26461</td>\n",
       "<td>9</td>\n",
       "<td>24344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.5303</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.98145</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0040016</td>\n",
       "<td>b''</td>\n",
       "<td>20</td>\n",
       "<td>0</td>\n",
       "<td>230</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.406</td>\n",
       "<td>9</td>\n",
       "<td>2.6504</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018349</td>\n",
       "<td>9829</td>\n",
       "<td>15.25</td>\n",
       "<td>2851.81</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>2.0234</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "VariantChunkedTable(9643193, nbytes=3.7G, cbytes=424.9M, cratio=8.8, data=h5py._hl.group.Group)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vt = allel.VariantChunkedTable(callset['3L/variants'])\n",
    "vt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChunkedArray((9643193,), |S12, chunks=(10922,), nbytes=110.4M, cbytes=578.6K, cratio=195.3, compression=gzip, compression_opts=3, data=h5py._hl.dataset.Dataset)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vt['CHROM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>VariantChunkedTable(9643193, nbytes=147.1M, cbytes=13.4M, cratio=11.0, data=builtins.list)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>CHROM</th>\n",
       "<th>POS</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9790</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9798</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9812</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9818</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9829</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "VariantChunkedTable(9643193, nbytes=147.1M, cbytes=13.4M, cratio=11.0, data=builtins.list)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vt[['CHROM', 'POS']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "query1 = '(AN == 1530) & (MQ > 40)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 loop, best of 3: 471 ms per loop\n"
     ]
    }
   ],
   "source": [
    "%timeit vt.eval(query1, vm='numexpr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 loop, best of 3: 506 ms per loop\n"
     ]
    }
   ],
   "source": [
    "%timeit vt.eval(query1, vm='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((9643193,), 9249396)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cond = vt.eval(query1, vm='numexpr')\n",
    "cond.shape, cond.count_nonzero()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# can only be evaluated with python vm\n",
    "query2 = '(AN == 1530) & (MQ > 40) & (AC[:, 1] > 5)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.26 s, sys: 92 ms, total: 2.36 s\n",
      "Wall time: 1.3 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ChunkedArray((9643193,), bool, chunks=(1048576,), nbytes=9.2M, cbytes=1.9M, cratio=4.9, compression=blosc, compression_opts={'clevel': 5, 'shuffle': 1, 'cname': 'blosclz'}, data=zarr.core.Array)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time vt.eval(query2, vm='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 loops, best of 3: 14.7 ms per loop\n"
     ]
    }
   ],
   "source": [
    "vt_bcolzmem = vt.copy(storage='bcolzmem', stop=1000000)\n",
    "%timeit vt_bcolzmem.eval(query1, vm='numexpr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.44 s, sys: 276 ms, total: 2.72 s\n",
      "Wall time: 881 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>VariantChunkedTable(950097, nbytes=368.8M, cbytes=63.7M, cratio=5.8, data=allel.chunked.storage_zarr.ZarrTable)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>ABHet</th>\n",
       "<th>ABHom</th>\n",
       "<th>AC</th>\n",
       "<th>AF</th>\n",
       "<th>ALT</th>\n",
       "<th>AN</th>\n",
       "<th>ANN</th>\n",
       "<th>Accessible</th>\n",
       "<th>BaseCounts</th>\n",
       "<th>BaseQRankSum</th>\n",
       "<th>CHROM</th>\n",
       "<th>Coverage</th>\n",
       "<th>CoverageMQ0</th>\n",
       "<th>DP</th>\n",
       "<th>DS</th>\n",
       "<th>Dels</th>\n",
       "<th>FILTER_FS</th>\n",
       "<th>FILTER_HRun</th>\n",
       "<th>FILTER_HighCoverage</th>\n",
       "<th>FILTER_HighMQ0</th>\n",
       "<th>FILTER_LowCoverage</th>\n",
       "<th>FILTER_LowMQ</th>\n",
       "<th>FILTER_LowQual</th>\n",
       "<th>FILTER_NoCoverage</th>\n",
       "<th>FILTER_PASS</th>\n",
       "<th>FILTER_QD</th>\n",
       "<th>FILTER_ReadPosRankSum</th>\n",
       "<th>FILTER_RefN</th>\n",
       "<th>FILTER_RepeatDUST</th>\n",
       "<th>FS</th>\n",
       "<th>HRun</th>\n",
       "<th>HW</th>\n",
       "<th>HaplotypeScore</th>\n",
       "<th>HighCoverage</th>\n",
       "<th>HighMQ0</th>\n",
       "<th>InbreedingCoeff</th>\n",
       "<th>LOF</th>\n",
       "<th>LowCoverage</th>\n",
       "<th>LowMQ</th>\n",
       "<th>LowPairing</th>\n",
       "<th>MLEAC</th>\n",
       "<th>MLEAF</th>\n",
       "<th>MQ</th>\n",
       "<th>MQ0</th>\n",
       "<th>MQRankSum</th>\n",
       "<th>NDA</th>\n",
       "<th>NMD</th>\n",
       "<th>NoCoverage</th>\n",
       "<th>OND</th>\n",
       "<th>POS</th>\n",
       "<th>QD</th>\n",
       "<th>QUAL</th>\n",
       "<th>REF</th>\n",
       "<th>RPA</th>\n",
       "<th>RU</th>\n",
       "<th>ReadPosRankSum</th>\n",
       "<th>RefMasked</th>\n",
       "<th>RefN</th>\n",
       "<th>RepeatDUST</th>\n",
       "<th>RepeatMasker</th>\n",
       "<th>RepeatTRF</th>\n",
       "<th>STR</th>\n",
       "<th>VariantType</th>\n",
       "<th>is_snp</th>\n",
       "<th>num_alleles</th>\n",
       "<th>svlen</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>0.50195</td>\n",
       "<td>0.99805</td>\n",
       "<td>[3 0 0]</td>\n",
       "<td>[ 0.00196075  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   21 25251    11    77]</td>\n",
       "<td>-1.8486</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25889</td>\n",
       "<td>41</td>\n",
       "<td>23344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.88184</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0332</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>-0.0025005</td>\n",
       "<td>b''</td>\n",
       "<td>18</td>\n",
       "<td>1</td>\n",
       "<td>368</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>42</td>\n",
       "<td>0.30811</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.002327</td>\n",
       "<td>9790</td>\n",
       "<td>14.047</td>\n",
       "<td>1447.12</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.22998</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[1 0 0]</td>\n",
       "<td>[ 0.00065374  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   39     9 25274    18]</td>\n",
       "<td>-3.5195</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25916</td>\n",
       "<td>14</td>\n",
       "<td>23321</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.4609</td>\n",
       "<td>0</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0449</td>\n",
       "<td>5</td>\n",
       "<td>0</td>\n",
       "<td>-0.0012999</td>\n",
       "<td>b''</td>\n",
       "<td>15</td>\n",
       "<td>1</td>\n",
       "<td>370</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>15</td>\n",
       "<td>3.2109</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017834</td>\n",
       "<td>9798</td>\n",
       "<td>13.781</td>\n",
       "<td>551.05</td>\n",
       "<td>b'G'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.13501</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.52783</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   45 25873    13    16]</td>\n",
       "<td>-3.0566</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26223</td>\n",
       "<td>8</td>\n",
       "<td>23877</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.3818</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0645</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>23</td>\n",
       "<td>0</td>\n",
       "<td>300</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.094</td>\n",
       "<td>8</td>\n",
       "<td>2.498</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018845</td>\n",
       "<td>9812</td>\n",
       "<td>13.273</td>\n",
       "<td>729.69</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.077026</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   43    21     9 26227]</td>\n",
       "<td>0.18298</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26516</td>\n",
       "<td>8</td>\n",
       "<td>24217</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.94775</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>19</td>\n",
       "<td>0</td>\n",
       "<td>279</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.156</td>\n",
       "<td>8</td>\n",
       "<td>3.2871</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017052</td>\n",
       "<td>9818</td>\n",
       "<td>14.352</td>\n",
       "<td>1147.7</td>\n",
       "<td>b'T'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.68213</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.48804</td>\n",
       "<td>0.99805</td>\n",
       "<td>[4 0 0]</td>\n",
       "<td>[ 0.00261307  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   14 26351     2   129]</td>\n",
       "<td>-3.1367</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26461</td>\n",
       "<td>9</td>\n",
       "<td>24344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.5303</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.98145</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0040016</td>\n",
       "<td>b''</td>\n",
       "<td>20</td>\n",
       "<td>0</td>\n",
       "<td>230</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.406</td>\n",
       "<td>9</td>\n",
       "<td>2.6504</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018349</td>\n",
       "<td>9829</td>\n",
       "<td>15.25</td>\n",
       "<td>2851.81</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>2.0234</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "VariantChunkedTable(950097, nbytes=368.8M, cbytes=63.7M, cratio=5.8, data=allel.chunked.storage_zarr.ZarrTable)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time vt_bcolzmem.query(query1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 loops, best of 3: 20.9 ms per loop\n"
     ]
    }
   ],
   "source": [
    "vt_hdf5mem_zlib1 = vt.copy(storage='hdf5mem_zlib1', stop=1000000)\n",
    "%timeit vt_hdf5mem_zlib1.eval(query1, vm='numexpr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.94 s, sys: 148 ms, total: 2.09 s\n",
      "Wall time: 1.32 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>VariantChunkedTable(950097, nbytes=368.8M, cbytes=63.7M, cratio=5.8, data=allel.chunked.storage_zarr.ZarrTable)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>ABHet</th>\n",
       "<th>ABHom</th>\n",
       "<th>AC</th>\n",
       "<th>AF</th>\n",
       "<th>ALT</th>\n",
       "<th>AN</th>\n",
       "<th>ANN</th>\n",
       "<th>Accessible</th>\n",
       "<th>BaseCounts</th>\n",
       "<th>BaseQRankSum</th>\n",
       "<th>CHROM</th>\n",
       "<th>Coverage</th>\n",
       "<th>CoverageMQ0</th>\n",
       "<th>DP</th>\n",
       "<th>DS</th>\n",
       "<th>Dels</th>\n",
       "<th>FILTER_FS</th>\n",
       "<th>FILTER_HRun</th>\n",
       "<th>FILTER_HighCoverage</th>\n",
       "<th>FILTER_HighMQ0</th>\n",
       "<th>FILTER_LowCoverage</th>\n",
       "<th>FILTER_LowMQ</th>\n",
       "<th>FILTER_LowQual</th>\n",
       "<th>FILTER_NoCoverage</th>\n",
       "<th>FILTER_PASS</th>\n",
       "<th>FILTER_QD</th>\n",
       "<th>FILTER_ReadPosRankSum</th>\n",
       "<th>FILTER_RefN</th>\n",
       "<th>FILTER_RepeatDUST</th>\n",
       "<th>FS</th>\n",
       "<th>HRun</th>\n",
       "<th>HW</th>\n",
       "<th>HaplotypeScore</th>\n",
       "<th>HighCoverage</th>\n",
       "<th>HighMQ0</th>\n",
       "<th>InbreedingCoeff</th>\n",
       "<th>LOF</th>\n",
       "<th>LowCoverage</th>\n",
       "<th>LowMQ</th>\n",
       "<th>LowPairing</th>\n",
       "<th>MLEAC</th>\n",
       "<th>MLEAF</th>\n",
       "<th>MQ</th>\n",
       "<th>MQ0</th>\n",
       "<th>MQRankSum</th>\n",
       "<th>NDA</th>\n",
       "<th>NMD</th>\n",
       "<th>NoCoverage</th>\n",
       "<th>OND</th>\n",
       "<th>POS</th>\n",
       "<th>QD</th>\n",
       "<th>QUAL</th>\n",
       "<th>REF</th>\n",
       "<th>RPA</th>\n",
       "<th>RU</th>\n",
       "<th>ReadPosRankSum</th>\n",
       "<th>RefMasked</th>\n",
       "<th>RefN</th>\n",
       "<th>RepeatDUST</th>\n",
       "<th>RepeatMasker</th>\n",
       "<th>RepeatTRF</th>\n",
       "<th>STR</th>\n",
       "<th>VariantType</th>\n",
       "<th>is_snp</th>\n",
       "<th>num_alleles</th>\n",
       "<th>svlen</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>0.50195</td>\n",
       "<td>0.99805</td>\n",
       "<td>[3 0 0]</td>\n",
       "<td>[ 0.00196075  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   21 25251    11    77]</td>\n",
       "<td>-1.8486</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25889</td>\n",
       "<td>41</td>\n",
       "<td>23344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.88184</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0332</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>-0.0025005</td>\n",
       "<td>b''</td>\n",
       "<td>18</td>\n",
       "<td>1</td>\n",
       "<td>368</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>42</td>\n",
       "<td>0.30811</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.002327</td>\n",
       "<td>9790</td>\n",
       "<td>14.047</td>\n",
       "<td>1447.12</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.22998</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[1 0 0]</td>\n",
       "<td>[ 0.00065374  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   39     9 25274    18]</td>\n",
       "<td>-3.5195</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25916</td>\n",
       "<td>14</td>\n",
       "<td>23321</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.4609</td>\n",
       "<td>0</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0449</td>\n",
       "<td>5</td>\n",
       "<td>0</td>\n",
       "<td>-0.0012999</td>\n",
       "<td>b''</td>\n",
       "<td>15</td>\n",
       "<td>1</td>\n",
       "<td>370</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>15</td>\n",
       "<td>3.2109</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017834</td>\n",
       "<td>9798</td>\n",
       "<td>13.781</td>\n",
       "<td>551.05</td>\n",
       "<td>b'G'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.13501</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.52783</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   45 25873    13    16]</td>\n",
       "<td>-3.0566</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26223</td>\n",
       "<td>8</td>\n",
       "<td>23877</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.3818</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0645</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>23</td>\n",
       "<td>0</td>\n",
       "<td>300</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.094</td>\n",
       "<td>8</td>\n",
       "<td>2.498</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018845</td>\n",
       "<td>9812</td>\n",
       "<td>13.273</td>\n",
       "<td>729.69</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.077026</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   43    21     9 26227]</td>\n",
       "<td>0.18298</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26516</td>\n",
       "<td>8</td>\n",
       "<td>24217</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.94775</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>19</td>\n",
       "<td>0</td>\n",
       "<td>279</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.156</td>\n",
       "<td>8</td>\n",
       "<td>3.2871</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017052</td>\n",
       "<td>9818</td>\n",
       "<td>14.352</td>\n",
       "<td>1147.7</td>\n",
       "<td>b'T'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.68213</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.48804</td>\n",
       "<td>0.99805</td>\n",
       "<td>[4 0 0]</td>\n",
       "<td>[ 0.00261307  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   14 26351     2   129]</td>\n",
       "<td>-3.1367</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26461</td>\n",
       "<td>9</td>\n",
       "<td>24344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.5303</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.98145</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0040016</td>\n",
       "<td>b''</td>\n",
       "<td>20</td>\n",
       "<td>0</td>\n",
       "<td>230</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.406</td>\n",
       "<td>9</td>\n",
       "<td>2.6504</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018349</td>\n",
       "<td>9829</td>\n",
       "<td>15.25</td>\n",
       "<td>2851.81</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>2.0234</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "VariantChunkedTable(950097, nbytes=368.8M, cbytes=63.7M, cratio=5.8, data=allel.chunked.storage_zarr.ZarrTable)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%time vt_hdf5mem_zlib1.query(query1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 loops, best of 3: 6.45 ms per loop\n"
     ]
    }
   ],
   "source": [
    "vt_zarrmem = vt.copy(storage='zarrmem', stop=1000000)\n",
    "%timeit vt_zarrmem.eval(query1, vm='numexpr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>VariantChunkedTable(1000000, nbytes=388.1M, cbytes=67.1M, cratio=5.8, data=allel.chunked.storage_zarr.ZarrTable)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>ABHet</th>\n",
       "<th>ABHom</th>\n",
       "<th>AC</th>\n",
       "<th>AF</th>\n",
       "<th>ALT</th>\n",
       "<th>AN</th>\n",
       "<th>ANN</th>\n",
       "<th>Accessible</th>\n",
       "<th>BaseCounts</th>\n",
       "<th>BaseQRankSum</th>\n",
       "<th>CHROM</th>\n",
       "<th>Coverage</th>\n",
       "<th>CoverageMQ0</th>\n",
       "<th>DP</th>\n",
       "<th>DS</th>\n",
       "<th>Dels</th>\n",
       "<th>FILTER_FS</th>\n",
       "<th>FILTER_HRun</th>\n",
       "<th>FILTER_HighCoverage</th>\n",
       "<th>FILTER_HighMQ0</th>\n",
       "<th>FILTER_LowCoverage</th>\n",
       "<th>FILTER_LowMQ</th>\n",
       "<th>FILTER_LowQual</th>\n",
       "<th>FILTER_NoCoverage</th>\n",
       "<th>FILTER_PASS</th>\n",
       "<th>FILTER_QD</th>\n",
       "<th>FILTER_ReadPosRankSum</th>\n",
       "<th>FILTER_RefN</th>\n",
       "<th>FILTER_RepeatDUST</th>\n",
       "<th>FS</th>\n",
       "<th>HRun</th>\n",
       "<th>HW</th>\n",
       "<th>HaplotypeScore</th>\n",
       "<th>HighCoverage</th>\n",
       "<th>HighMQ0</th>\n",
       "<th>InbreedingCoeff</th>\n",
       "<th>LOF</th>\n",
       "<th>LowCoverage</th>\n",
       "<th>LowMQ</th>\n",
       "<th>LowPairing</th>\n",
       "<th>MLEAC</th>\n",
       "<th>MLEAF</th>\n",
       "<th>MQ</th>\n",
       "<th>MQ0</th>\n",
       "<th>MQRankSum</th>\n",
       "<th>NDA</th>\n",
       "<th>NMD</th>\n",
       "<th>NoCoverage</th>\n",
       "<th>OND</th>\n",
       "<th>POS</th>\n",
       "<th>QD</th>\n",
       "<th>QUAL</th>\n",
       "<th>REF</th>\n",
       "<th>RPA</th>\n",
       "<th>RU</th>\n",
       "<th>ReadPosRankSum</th>\n",
       "<th>RefMasked</th>\n",
       "<th>RefN</th>\n",
       "<th>RepeatDUST</th>\n",
       "<th>RepeatMasker</th>\n",
       "<th>RepeatTRF</th>\n",
       "<th>STR</th>\n",
       "<th>VariantType</th>\n",
       "<th>is_snp</th>\n",
       "<th>num_alleles</th>\n",
       "<th>svlen</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>0.50195</td>\n",
       "<td>0.99805</td>\n",
       "<td>[3 0 0]</td>\n",
       "<td>[ 0.00196075  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   21 25251    11    77]</td>\n",
       "<td>-1.8486</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25889</td>\n",
       "<td>41</td>\n",
       "<td>23344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.88184</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0332</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>-0.0025005</td>\n",
       "<td>b''</td>\n",
       "<td>18</td>\n",
       "<td>1</td>\n",
       "<td>368</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>42</td>\n",
       "<td>0.30811</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.002327</td>\n",
       "<td>9790</td>\n",
       "<td>14.047</td>\n",
       "<td>1447.12</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.22998</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[1 0 0]</td>\n",
       "<td>[ 0.00065374  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   39     9 25274    18]</td>\n",
       "<td>-3.5195</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25916</td>\n",
       "<td>14</td>\n",
       "<td>23321</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.4609</td>\n",
       "<td>0</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0449</td>\n",
       "<td>5</td>\n",
       "<td>0</td>\n",
       "<td>-0.0012999</td>\n",
       "<td>b''</td>\n",
       "<td>15</td>\n",
       "<td>1</td>\n",
       "<td>370</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>15</td>\n",
       "<td>3.2109</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017834</td>\n",
       "<td>9798</td>\n",
       "<td>13.781</td>\n",
       "<td>551.05</td>\n",
       "<td>b'G'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.13501</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.52783</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   45 25873    13    16]</td>\n",
       "<td>-3.0566</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26223</td>\n",
       "<td>8</td>\n",
       "<td>23877</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.3818</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0645</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>23</td>\n",
       "<td>0</td>\n",
       "<td>300</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.094</td>\n",
       "<td>8</td>\n",
       "<td>2.498</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018845</td>\n",
       "<td>9812</td>\n",
       "<td>13.273</td>\n",
       "<td>729.69</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.077026</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   43    21     9 26227]</td>\n",
       "<td>0.18298</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26516</td>\n",
       "<td>8</td>\n",
       "<td>24217</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.94775</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>19</td>\n",
       "<td>0</td>\n",
       "<td>279</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.156</td>\n",
       "<td>8</td>\n",
       "<td>3.2871</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017052</td>\n",
       "<td>9818</td>\n",
       "<td>14.352</td>\n",
       "<td>1147.7</td>\n",
       "<td>b'T'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.68213</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.48804</td>\n",
       "<td>0.99805</td>\n",
       "<td>[4 0 0]</td>\n",
       "<td>[ 0.00261307  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   14 26351     2   129]</td>\n",
       "<td>-3.1367</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26461</td>\n",
       "<td>9</td>\n",
       "<td>24344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.5303</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.98145</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0040016</td>\n",
       "<td>b''</td>\n",
       "<td>20</td>\n",
       "<td>0</td>\n",
       "<td>230</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.406</td>\n",
       "<td>9</td>\n",
       "<td>2.6504</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018349</td>\n",
       "<td>9829</td>\n",
       "<td>15.25</td>\n",
       "<td>2851.81</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>2.0234</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "VariantChunkedTable(1000000, nbytes=388.1M, cbytes=67.1M, cratio=5.8, data=allel.chunked.storage_zarr.ZarrTable)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vt_zarrmem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>VariantTable((10,), dtype=[('ABHet', '&lt;f2'), ('ABHom', '&lt;f2'), ('AC', '&lt;u2', (3,)), ('AF', '&lt;f2', (3,)), ('ALT', 'S1', (3,)), ('AN', '&lt;u2'), ('ANN', [('Allele', 'S12'), ('Annotation', 'S34'), ('Annotation_Impact', 'S8'), ('Gene_Name', 'S14'), ('Gene_ID', 'S14'), ('Feature_Type', 'S20'), ('Feature_ID', 'S14'), ('Transcript_BioType', 'S20'), ('Rank', 'i1'), ('HGVS_c', 'S12'), ('HGVS_p', 'S14'), ('cDNA_pos', '&lt;i4'), ('cDNA_length', '&lt;i4'), ('CDS_pos', '&lt;i4'), ('CDS_length', '&lt;i4'), ('AA_pos', '&lt;i4'), ('AA_length', '&lt;i4'), ('Distance', '&lt;i4')]), ('Accessible', '?'), ('BaseCounts', '&lt;i4', (4,)), ('BaseQRankSum', '&lt;f2'), ('CHROM', 'S12'), ('Coverage', '&lt;i4'), ('CoverageMQ0', '&lt;i4'), ('DP', '&lt;i4'), ('DS', '?'), ('Dels', '&lt;f2'), ('FILTER_FS', '?'), ('FILTER_HRun', '?'), ('FILTER_HighCoverage', '?'), ('FILTER_HighMQ0', '?'), ('FILTER_LowCoverage', '?'), ('FILTER_LowMQ', '?'), ('FILTER_LowQual', '?'), ('FILTER_NoCoverage', '?'), ('FILTER_PASS', '?'), ('FILTER_QD', '?'), ('FILTER_ReadPosRankSum', '?'), ('FILTER_RefN', '?'), ('FILTER_RepeatDUST', '?'), ('FS', '&lt;f2'), ('HRun', 'u1'), ('HW', '&lt;f4'), ('HaplotypeScore', '&lt;f2'), ('HighCoverage', '&lt;i4'), ('HighMQ0', '&lt;i4'), ('InbreedingCoeff', '&lt;f2'), ('LOF', 'S12'), ('LowCoverage', '&lt;i4'), ('LowMQ', '&lt;i4'), ('LowPairing', '&lt;i4'), ('MLEAC', '&lt;u2', (3,)), ('MLEAF', '&lt;f2', (3,)), ('MQ', '&lt;f2'), ('MQ0', '&lt;i4'), ('MQRankSum', '&lt;f2'), ('NDA', '&lt;i4'), ('NMD', 'S12'), ('NoCoverage', '&lt;i4'), ('OND', '&lt;f2'), ('POS', '&lt;i4'), ('QD', '&lt;f2'), ('QUAL', '&lt;f4'), ('REF', 'S1'), ('RPA', '&lt;u2'), ('RU', 'S12'), ('ReadPosRankSum', '&lt;f2'), ('RefMasked', '?'), ('RefN', '?'), ('RepeatDUST', '?'), ('RepeatMasker', '?'), ('RepeatTRF', '?'), ('STR', '?'), ('VariantType', 'S12'), ('is_snp', '?'), ('num_alleles', 'u1'), ('svlen', '&lt;i4')])</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>ABHet</th>\n",
       "<th>ABHom</th>\n",
       "<th>AC</th>\n",
       "<th>AF</th>\n",
       "<th>ALT</th>\n",
       "<th>AN</th>\n",
       "<th>ANN</th>\n",
       "<th>Accessible</th>\n",
       "<th>BaseCounts</th>\n",
       "<th>BaseQRankSum</th>\n",
       "<th>CHROM</th>\n",
       "<th>Coverage</th>\n",
       "<th>CoverageMQ0</th>\n",
       "<th>DP</th>\n",
       "<th>DS</th>\n",
       "<th>Dels</th>\n",
       "<th>FILTER_FS</th>\n",
       "<th>FILTER_HRun</th>\n",
       "<th>FILTER_HighCoverage</th>\n",
       "<th>FILTER_HighMQ0</th>\n",
       "<th>FILTER_LowCoverage</th>\n",
       "<th>FILTER_LowMQ</th>\n",
       "<th>FILTER_LowQual</th>\n",
       "<th>FILTER_NoCoverage</th>\n",
       "<th>FILTER_PASS</th>\n",
       "<th>FILTER_QD</th>\n",
       "<th>FILTER_ReadPosRankSum</th>\n",
       "<th>FILTER_RefN</th>\n",
       "<th>FILTER_RepeatDUST</th>\n",
       "<th>FS</th>\n",
       "<th>HRun</th>\n",
       "<th>HW</th>\n",
       "<th>HaplotypeScore</th>\n",
       "<th>HighCoverage</th>\n",
       "<th>HighMQ0</th>\n",
       "<th>InbreedingCoeff</th>\n",
       "<th>LOF</th>\n",
       "<th>LowCoverage</th>\n",
       "<th>LowMQ</th>\n",
       "<th>LowPairing</th>\n",
       "<th>MLEAC</th>\n",
       "<th>MLEAF</th>\n",
       "<th>MQ</th>\n",
       "<th>MQ0</th>\n",
       "<th>MQRankSum</th>\n",
       "<th>NDA</th>\n",
       "<th>NMD</th>\n",
       "<th>NoCoverage</th>\n",
       "<th>OND</th>\n",
       "<th>POS</th>\n",
       "<th>QD</th>\n",
       "<th>QUAL</th>\n",
       "<th>REF</th>\n",
       "<th>RPA</th>\n",
       "<th>RU</th>\n",
       "<th>ReadPosRankSum</th>\n",
       "<th>RefMasked</th>\n",
       "<th>RefN</th>\n",
       "<th>RepeatDUST</th>\n",
       "<th>RepeatMasker</th>\n",
       "<th>RepeatTRF</th>\n",
       "<th>STR</th>\n",
       "<th>VariantType</th>\n",
       "<th>is_snp</th>\n",
       "<th>num_alleles</th>\n",
       "<th>svlen</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>0.50195</td>\n",
       "<td>0.99805</td>\n",
       "<td>[3 0 0]</td>\n",
       "<td>[ 0.00196075  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   21 25251    11    77]</td>\n",
       "<td>-1.8486</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25889</td>\n",
       "<td>41</td>\n",
       "<td>23344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.88184</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0332</td>\n",
       "<td>3</td>\n",
       "<td>0</td>\n",
       "<td>-0.0025005</td>\n",
       "<td>b''</td>\n",
       "<td>18</td>\n",
       "<td>1</td>\n",
       "<td>368</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>42</td>\n",
       "<td>0.30811</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.002327</td>\n",
       "<td>9790</td>\n",
       "<td>14.047</td>\n",
       "<td>1447.12</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.22998</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[1 0 0]</td>\n",
       "<td>[ 0.00065374  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   39     9 25274    18]</td>\n",
       "<td>-3.5195</td>\n",
       "<td>b'3L'</td>\n",
       "<td>25916</td>\n",
       "<td>14</td>\n",
       "<td>23321</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.4609</td>\n",
       "<td>0</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0449</td>\n",
       "<td>5</td>\n",
       "<td>0</td>\n",
       "<td>-0.0012999</td>\n",
       "<td>b''</td>\n",
       "<td>15</td>\n",
       "<td>1</td>\n",
       "<td>370</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>54.844</td>\n",
       "<td>15</td>\n",
       "<td>3.2109</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017834</td>\n",
       "<td>9798</td>\n",
       "<td>13.781</td>\n",
       "<td>551.05</td>\n",
       "<td>b'G'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.13501</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.52783</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   45 25873    13    16]</td>\n",
       "<td>-3.0566</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26223</td>\n",
       "<td>8</td>\n",
       "<td>23877</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.3818</td>\n",
       "<td>1</td>\n",
       "<td>0.0</td>\n",
       "<td>1.0645</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>23</td>\n",
       "<td>0</td>\n",
       "<td>300</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.094</td>\n",
       "<td>8</td>\n",
       "<td>2.498</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018845</td>\n",
       "<td>9812</td>\n",
       "<td>13.273</td>\n",
       "<td>729.69</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>-0.077026</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.5249</td>\n",
       "<td>0.99805</td>\n",
       "<td>[2 0 0]</td>\n",
       "<td>[ 0.00130653  0.          0.        ]</td>\n",
       "<td>[b'A' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   43    21     9 26227]</td>\n",
       "<td>0.18298</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26516</td>\n",
       "<td>8</td>\n",
       "<td>24217</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.94775</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0017004</td>\n",
       "<td>b''</td>\n",
       "<td>19</td>\n",
       "<td>0</td>\n",
       "<td>279</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.156</td>\n",
       "<td>8</td>\n",
       "<td>3.2871</td>\n",
       "<td>2</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0017052</td>\n",
       "<td>9818</td>\n",
       "<td>14.352</td>\n",
       "<td>1147.7</td>\n",
       "<td>b'T'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>0.68213</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>0.48804</td>\n",
       "<td>0.99805</td>\n",
       "<td>[4 0 0]</td>\n",
       "<td>[ 0.00261307  0.          0.        ]</td>\n",
       "<td>[b'T' b'' b'']</td>\n",
       "<td>1530</td>\n",
       "<td>(b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1)</td>\n",
       "<td>True</td>\n",
       "<td>[   14 26351     2   129]</td>\n",
       "<td>-3.1367</td>\n",
       "<td>b'3L'</td>\n",
       "<td>26461</td>\n",
       "<td>9</td>\n",
       "<td>24344</td>\n",
       "<td>False</td>\n",
       "<td>0.0</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1.5303</td>\n",
       "<td>2</td>\n",
       "<td>0.0</td>\n",
       "<td>0.98145</td>\n",
       "<td>7</td>\n",
       "<td>0</td>\n",
       "<td>-0.0040016</td>\n",
       "<td>b''</td>\n",
       "<td>20</td>\n",
       "<td>0</td>\n",
       "<td>230</td>\n",
       "<td>[0 0 0]</td>\n",
       "<td>[ 0.  0.  0.]</td>\n",
       "<td>55.406</td>\n",
       "<td>9</td>\n",
       "<td>2.6504</td>\n",
       "<td>1</td>\n",
       "<td>b''</td>\n",
       "<td>0</td>\n",
       "<td>0.0018349</td>\n",
       "<td>9829</td>\n",
       "<td>15.25</td>\n",
       "<td>2851.81</td>\n",
       "<td>b'C'</td>\n",
       "<td>0</td>\n",
       "<td>b''</td>\n",
       "<td>2.0234</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>2</td>\n",
       "<td>0</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "VariantTable((10,), dtype=[('ABHet', '<f2'), ('ABHom', '<f2'), ('AC', '<u2', (3,)), ('AF', '<f2', (3,)), ('ALT', 'S1', (3,)), ('AN', '<u2'), ('ANN', [('Allele', 'S12'), ('Annotation', 'S34'), ('Annotation_Impact', 'S8'), ('Gene_Name', 'S14'), ('Gene_ID', 'S14'), ('Feature_Type', 'S20'), ('Feature_ID', 'S14'), ('Transcript_BioType', 'S20'), ('Rank', 'i1'), ('HGVS_c', 'S12'), ('HGVS_p', 'S14'), ('cDNA_pos', '<i4'), ('cDNA_length', '<i4'), ('CDS_pos', '<i4'), ('CDS_length', '<i4'), ('AA_pos', '<i4'), ('AA_length', '<i4'), ('Distance', '<i4')]), ('Accessible', '?'), ('BaseCounts', '<i4', (4,)), ('BaseQRankSum', '<f2'), ('CHROM', 'S12'), ('Coverage', '<i4'), ('CoverageMQ0', '<i4'), ('DP', '<i4'), ('DS', '?'), ('Dels', '<f2'), ('FILTER_FS', '?'), ('FILTER_HRun', '?'), ('FILTER_HighCoverage', '?'), ('FILTER_HighMQ0', '?'), ('FILTER_LowCoverage', '?'), ('FILTER_LowMQ', '?'), ('FILTER_LowQual', '?'), ('FILTER_NoCoverage', '?'), ('FILTER_PASS', '?'), ('FILTER_QD', '?'), ('FILTER_ReadPosRankSum', '?'), ('FILTER_RefN', '?'), ('FILTER_RepeatDUST', '?'), ('FS', '<f2'), ('HRun', 'u1'), ('HW', '<f4'), ('HaplotypeScore', '<f2'), ('HighCoverage', '<i4'), ('HighMQ0', '<i4'), ('InbreedingCoeff', '<f2'), ('LOF', 'S12'), ('LowCoverage', '<i4'), ('LowMQ', '<i4'), ('LowPairing', '<i4'), ('MLEAC', '<u2', (3,)), ('MLEAF', '<f2', (3,)), ('MQ', '<f2'), ('MQ0', '<i4'), ('MQRankSum', '<f2'), ('NDA', '<i4'), ('NMD', 'S12'), ('NoCoverage', '<i4'), ('OND', '<f2'), ('POS', '<i4'), ('QD', '<f2'), ('QUAL', '<f4'), ('REF', 'S1'), ('RPA', '<u2'), ('RU', 'S12'), ('ReadPosRankSum', '<f2'), ('RefMasked', '?'), ('RefN', '?'), ('RepeatDUST', '?'), ('RepeatMasker', '?'), ('RepeatTRF', '?'), ('STR', '?'), ('VariantType', 'S12'), ('is_snp', '?'), ('num_alleles', 'u1'), ('svlen', '<i4')])\n",
       "[ (0.501953125, 0.998046875, [3, 0, 0], [0.00196075439453125, 0.0, 0.0], [b'T', b'', b''], 1530, (b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [21, 25251, 11, 77], -1.8486328125, b'3L', 25889, 41, 23344, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 0.8818359375, 1, 0.0, 1.033203125, 3, 0, -0.0025005340576171875, b'', 18, 1, 368, [0, 0, 0], [0.0, 0.0, 0.0], 54.84375, 42, 0.30810546875, 1, b'', 0, 0.00232696533203125, 9790, 14.046875, 1447.1199951171875, b'C', 0, b'', 0.22998046875, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.52490234375, 0.998046875, [1, 0, 0], [0.0006537437438964844, 0.0, 0.0], [b'A', b'', b''], 1530, (b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [39, 9, 25274, 18], -3.51953125, b'3L', 25916, 14, 23321, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 1.4609375, 0, 0.0, 1.044921875, 5, 0, -0.0012998580932617188, b'', 15, 1, 370, [0, 0, 0], [0.0, 0.0, 0.0], 54.84375, 15, 3.2109375, 2, b'', 0, 0.0017833709716796875, 9798, 13.78125, 551.0499877929688, b'G', 0, b'', -0.135009765625, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.52783203125, 0.998046875, [2, 0, 0], [0.0013065338134765625, 0.0, 0.0], [b'A', b'', b''], 1530, (b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [45, 25873, 13, 16], -3.056640625, b'3L', 26223, 8, 23877, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 1.3818359375, 1, 0.0, 1.064453125, 7, 0, -0.0017004013061523438, b'', 23, 0, 300, [0, 0, 0], [0.0, 0.0, 0.0], 55.09375, 8, 2.498046875, 1, b'', 0, 0.00188446044921875, 9812, 13.2734375, 729.6900024414062, b'C', 0, b'', -0.0770263671875, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.52490234375, 0.998046875, [2, 0, 0], [0.0013065338134765625, 0.0, 0.0], [b'A', b'', b''], 1530, (b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [43, 21, 9, 26227], 0.1829833984375, b'3L', 26516, 8, 24217, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 0.0, 2, 0.0, 0.94775390625, 7, 0, -0.0017004013061523438, b'', 19, 0, 279, [0, 0, 0], [0.0, 0.0, 0.0], 55.15625, 8, 3.287109375, 2, b'', 0, 0.001705169677734375, 9818, 14.3515625, 1147.699951171875, b'T', 0, b'', 0.68212890625, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.488037109375, 0.998046875, [4, 0, 0], [0.002613067626953125, 0.0, 0.0], [b'T', b'', b''], 1530, (b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [14, 26351, 2, 129], -3.13671875, b'3L', 26461, 9, 24344, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 1.5302734375, 2, 0.0, 0.9814453125, 7, 0, -0.004001617431640625, b'', 20, 0, 230, [0, 0, 0], [0.0, 0.0, 0.0], 55.40625, 9, 2.650390625, 1, b'', 0, 0.001834869384765625, 9829, 15.25, 2851.81005859375, b'C', 0, b'', 2.0234375, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.360107421875, 0.998046875, [1, 0, 0], [0.0006537437438964844, 0.0, 0.0], [b'A', b'', b''], 1530, (b'A', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [36, 13, 26322, 15], -3.634765625, b'3L', 26313, 9, 24258, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 2.33203125, 0, 0.0, 0.9853515625, 7, 0, -0.0010995864868164062, b'', 19, 0, 214, [0, 0, 0], [0.0, 0.0, 0.0], 55.5, 9, 2.451171875, 1, b'', 0, 0.0019931793212890625, 9831, 18.40625, 460.04998779296875, b'G', 0, b'', -0.615234375, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.5, 0.998046875, [1, 0, 0], [0.0006537437438964844, 0.0, 0.0], [b'T', b'', b''], 1530, (b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [31, 11, 26499, 30], -2.40625, b'3L', 26188, 8, 24437, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 0.0, 1, 0.0, 0.80908203125, 5, 0, -0.0010995864868164062, b'', 23, 0, 132, [0, 0, 0], [0.0, 0.0, 0.0], 56.0, 9, 4.68359375, 2, b'', 0, 0.0025081634521484375, 9849, 13.171875, 237.0500030517578, b'G', 0, b'', -0.00800323486328125, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.5791015625, 0.9990234375, [7, 0, 0], [0.004573822021484375, 0.0, 0.0], [b'C', b'', b''], 1530, (b'C', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [26769, 123, 2, 4], -2.650390625, b'3L', 26460, 8, 24779, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 3.34765625, 0, 0.0, 0.8349609375, 7, 0, -0.00550079345703125, b'', 22, 0, 118, [0, 0, 0], [0.0, 0.0, 0.0], 56.125, 9, 5.13671875, 1, b'', 0, 0.0008115768432617188, 9853, 11.5, 3012.679931640625, b'A', 0, b'', -0.08099365234375, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.462890625, 0.9990234375, [13, 0, 0], [0.0084991455078125, 0.0, 0.0], [b'C', b'', b''], 1530, (b'C', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [5, 259, 8, 27031], -2.638671875, b'3L', 26809, 8, 25159, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 4.22265625, 0, 60.0, 1.1494140625, 8, 0, 0.4189453125, b'', 22, 0, 101, [0, 0, 0], [0.0, 0.0, 0.0], 56.21875, 9, 7.3515625, 1, b'', 0, 0.0008540153503417969, 9859, 21.609375, 7994.509765625, b'T', 0, b'', 1.490234375, True, False, False, True, False, False, b'SNP', True, 2, 0)\n",
       " (0.47998046875, 0.998046875, [2, 0, 0], [0.0013065338134765625, 0.0, 0.0], [b'T', b'', b''], 1530, (b'T', b'intergenic_region', b'MODIFIER', b'AGAP010310', b'AGAP010310', b'intergenic_region', b'AGAP010310', b'.', -1, b'.', b'.', -1, -1, -1, -1, -1, -1, -1), True, [23, 27452, 11, 33], -2.70703125, b'3L', 26918, 7, 25350, False, 0.0, False, False, False, False, False, False, False, False, True, False, False, False, False, 0.7099609375, 0, 0.0, 1.3486328125, 7, 0, -0.0015001296997070312, b'', 19, 0, 78, [0, 0, 0], [0.0, 0.0, 0.0], 56.375, 8, 5.0, 1, b'', 0, 0.0018148422241210938, 9867, 12.890625, 644.6900024414062, b'C', 0, b'', -0.97412109375, True, False, False, True, False, False, b'SNP', True, 2, 0)]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vt_zarrmem[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Copy into existing HDF5 file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 group \"/variants\" (0 members)>"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5g = h5f.require_group('variants')\n",
    "h5g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for k in h5g:\n",
    "    del h5g[k]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<caption>VariantChunkedTable(100000, nbytes=2.3M, cbytes=300.8K, cratio=7.8, data=h5py._hl.group.Group)</caption>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>CHROM</th>\n",
       "<th>POS</th>\n",
       "<th>AN</th>\n",
       "<th>AC</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9790</td>\n",
       "<td>1530</td>\n",
       "<td>[3 0 0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9798</td>\n",
       "<td>1530</td>\n",
       "<td>[1 0 0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9812</td>\n",
       "<td>1530</td>\n",
       "<td>[2 0 0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9818</td>\n",
       "<td>1530</td>\n",
       "<td>[2 0 0]</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>b'3L'</td>\n",
       "<td>9829</td>\n",
       "<td>1530</td>\n",
       "<td>[4 0 0]</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "VariantChunkedTable(100000, nbytes=2.3M, cbytes=300.8K, cratio=7.8, data=h5py._hl.group.Group)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vtc = vt[['CHROM', 'POS', 'AN', 'AC']].copy(stop=100000, storage='hdf5_zlib1', group=h5g)\n",
    "vtc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 group \"/variants\" (4 members)>"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
